### ISSUE TITLE
Cannot Train with Depth Maps without Chamfer Distance

### ISSUE THREAD

**Issue by Wi-sc** [microsoft/SpareNet#13]
Hi, thanks for your released code. I'm  interested in the your methods that training with depth map. So I design a toy problem where I only train with depth map but I find points become NaN. Do you have any suggestions? Does this mean the depth maps cannot help training?

```
from p2i_utils import ComputeDepthMaps
compute_depth_maps = ComputeDepthMaps(projection="perspective", eyepos_scale=1.0, image_size=224).float()

def get_depth_render(points, requires_grad=True):
    depth_list = []
    for view_id in range(8):
        _depth = compute_depth_maps(points, view_id=view_id)
        depth_list.append(_depth)
    if requires_grad:
        return torch.cat(depth_list, dim=1)
    else:
        return torch.cat(depth_list, dim=1).detach()

trg_points = load_from_file()
input_points = torch.full([1, 2500, 3], 0.0, device=device, requires_grad=True)
optimizer = torch.optim.SGD([input_points ], lr=1.0, momentum=0.9)
Niter = 2000
loss_list = []

for i in range(Niter):
    optimizer.zero_grad()
    gt_depth = get_depth_render(trg_points, requires_grad=False)
    pred_depth = get_depth_render(input_points , requires_grad=True)
    loss_depth = torch.nn.L1Loss()(pred_depth, gt_depth)
    loss = loss_depth
    loss_list.append(loss.detach().cpu().item())
    loss.backward()
    optimizer.step()

```

**Comment by AlphaPav** [comment]
Hi, thanks for your interest! @Wi-sc I think Chamfer Distance is not the reason for NaN in your experiments. 

In our method, the supervision of depth maps can help optimize the point encoder and generator networks. However, in your case, you want depth maps to directly optimize the 3D Points themselves. It could be an interesting application of our point renderer. 
Could you decrease the learning rate? lr =1.0 might be too large.

**Comment by Wi-sc** [comment]
@AlphaPav Thanks for your reply. I have changed the learning rate into 0.01. But it also outputs NaN. And actually I also tried to try an auto-encoder network. But l1 distance between depth maps cannot regularize the training. I think this is not reasonable because Render4Completion actually do the experiments. Have you test the modified depth renderer under this setting?

**Comment by Wi-sc** [comment]
@AlphaPav Hi, I still cannot train a simple encoder-decoder network under supervision of depth maps.

**Comment by YANG-H** [comment]
Hi @Wi-sc , in my opinion, using depth maps as the only supervision without CD/EMD should not work.

Just imagine when only one `pred point` appears in the `pred_depth` and one `gt point` appears in the `gt_depth`, and there is a distance between the `gt point` and the `pred point`. When calculating the L1 loss between such two depth maps pixel-wisely, since all pixels nearby the `pred point` mostly see nothing but close-to-zero values in `gt_depth` (because the two points have a distance), then the gradients back-propagated to the `pred_depth` pixels can hardly reflect directional information concerning the location of the `gt point`. Therefore, the pixel-wise L1 loss will hardly make the two points closer.

Furthermore, in our `p2i` code, we implement the rendering with only a limited range around points for rendering efficiency: pixels far away from the rendered point are directly set to an arbitrary value. Under our implementation, it becomes more impossible that a L1 loss between depth-maps could ever work.
